src/kmerseek/sig2kmer.py

```diff
@@ -0,0 +1,176 @@
+# ! /usr/bin/env python3
+"""
+Given a signature file and a collection of sequences, output all of the
+k-mers and sequences that match a hashval in the signature file.
+
+Cribbed from https://github.com/dib-lab/sourmash/pull/724/
+"""
+
+from typing import Literal
+from tempfile import NamedTemporaryFile
+
+import sourmash
+import polars as pl
+import screed
+
+
+def _make_kmer_filename(sig):
+    return f"{sig}.kmers.pq"
+
+
+# TODO: maybe add this to sourmash.sig.__main__.kmers because it's useful to us,
+# could be useful to others
+def encode_protein(sequence, moltype):
+    """Convert protein sequence from 20-letter amino acid alphabet to degenerate alphabet"""
+    # Pre-determine the alphabet function based on moltype
+    if moltype == "hp":
+        alphabet_func = sourmash._lowlevel.lib.sourmash_aa_to_hp
+    elif moltype == "dayhoff":
+        alphabet_func = sourmash._lowlevel.lib.sourmash_aa_to_dayhoff
+    else:
+        raise ValueError(f"Unknown moltype: {moltype}")
+
+    # Convert the entire sequence to bytes once
+    byte_sequence = sequence.encode("utf-8")
+
+    # Apply the alphabet function to each byte in the sequence and join the result
+    degenerate = b"".join(
+        alphabet_func(letter.to_bytes(1, "big")) for letter in byte_sequence
+    ).decode()
+
+    return degenerate
+
+
+class Args:
+    """Dummy class to call what is normally a command line function from Python"""
+
+    def __init__(self, sig, fasta, moltype, ksize, scaled, save_kmers, save_sequences):
+        self.signatures = [sig]  # List of signature files
+        self.sequences = [fasta]  # List of sequence files
+        self.ksize = ksize
+        self.quiet = False
+        self.force = True
+        self.moltype = moltype
+        self.translate = False
+        self.check_sequence = True
+        self.save_kmers = save_kmers
+        self.save_sequences = save_sequences
+        self.picklist = None
+        self.scaled = scaled
+        self.dna = False
+        self.skipm1n3 = False
+        self.skipm2n3 = False
+        self.from_file = False
+
+        self.dayhoff = False
+        self.hp = False
+        self.protein = False
+        if moltype == "protein":
+            self.protein = True
+        elif moltype == "dayhoff":
+            self.dayhoff = True
+        elif moltype == "hp":
+            self.hp = True
+
+
+def add_encoding_to_kmers_pl(kmers, moltype):
+
+    # Apply encoding using a UDF (User Defined Function)
+    if moltype == "hp" or moltype == "dayhoff":
+        kmers_with_encoding = kmers.with_columns(
```

Comment:
Curious how UDF performance in polars compares to DuckDb and pyspark.  If you have this `kmers` dataframe without the encoding column saved to disk somewhere, I'd be interested in doing a little benchmarking